File: Week7/rkdtndk99/타겟 넘버.py
# BFS
from collections import deque

def solution(numbers, target):
    answer = 0
    queue = deque()
    queue.append([numbers[0], 0])
    queue.append([-1*numbers[0], 0])
    while queue:
        temp, idx = queue.popleft()
        idx += 1
        if idx < len(numbers):
            queue.append([temp+numbers[idx], idx])
            queue.append([temp-numbers[idx], idx])
        else :
            if temp == target:
                answer += 1
    return answer


# DFS
def solution(numbers, target):
    answer = 0

    def dfs(idx, result):
        stack = []
        if idx == len(numbers):
            if result == target:
                nonlocal answer
                answer += 1
            return
        else:
            dfs(idx + 1, result + numbers[idx])
            dfs(idx + 1, result - numbers[idx])

    dfs(0, 0)
    return answer

File: Week7/chaewoniii/타겟넘버.java
class Solution {
    public int answer;
    public int solution(int[] numbers, int target) {
        answer=0;
        dfs(numbers,target,0,0);
        return answer;
    }
    public void dfs(int[] numbers, int target, int index, int sum){
        //끝까지 탐색한 경우
        if(index==numbers.length){
            if(sum==target){
                answer++;
                return;
            }
            return;
        }
        dfs(numbers,target,index+1,sum+numbers[index]);
        dfs(numbers,target,index+1,sum-numbers[index]); 
    }
}
